packages/angular_devkit/schematics/src/engine/task.ts

```diff
@@ -0,0 +1,117 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+import { Observable } from 'rxjs/Observable';
+
+export interface TaskConfiguration<T = {}> {
+  name: string;
+  dependencies?: Array<TaskId>;
+  failureMode?: 'ignore' | 'warn' | 'error';
+  failureMessage?: string;
+  options?: T;
+}
+
+export interface TaskConfigurationGenerator<T = {}> {
+  toConfiguration(): TaskConfiguration<T>;
+}
+
+export type TaskExecutor<T = {}> = (options?: T) => Observable<void>;
+
+export interface TaskId {
+  readonly id: number;
+  readonly priority:  number;
+  readonly configuration: Readonly<TaskConfiguration>;
+}
+
+function compareTasks(x: TaskId, y: TaskId) {
+  if (x.priority === y.priority) {
+    return 0;
+  }
+
+  return x.priority > y.priority ? 1 : -1;
+}
+
+export class TaskScheduler {
+  private _queue = new PriorityQueue<TaskId>(compareTasks);
+  private static _taskIdCounter = 1;
+
+  private _calculatePriority(dependencies: Set<TaskId>): number {
+    if (dependencies.size === 0) {
+      return 0;
+    }
+
+    const prio = [...dependencies].reduce((prio, task) => prio + task.priority, 1);
+
+    return prio;
+  }
+
+  schedule<T>(taskConfiguration: TaskConfiguration<T>): TaskId {
+
+    const dependencies = new Set<TaskId>(taskConfiguration.dependencies);
+    const priority = this._calculatePriority(dependencies);
+
+    const task = {
+      id: TaskScheduler._taskIdCounter++,
+      priority,
+      configuration: taskConfiguration,
+    };
+
+    this._queue.push(task);
+
+    return task;
+  }
+
+  finalize(): Observable<TaskId> {
+    return Observable.from(this._queue.toArray());
+  }
+
+}
+
+// Naive priority queue
+class PriorityQueue<T> {
```

Comment:
Could you move this to `core`? Seems like something we could use somewhere else.

Reply:
Done.